Understanding TCP/IP Transport Layer (Layer 3) protocols:TCP and UDP (#1463)
section-engineering-education/engineering-education: articles/understanding-tcp-ip-transport-layer-protocols/index.md

# Understanding TCP/IP Transport Layer (Layer 4) protocols: TCP and UDP  

![Hero image](/engineering-education/understanding-tcp-ip-transport-layer-protocols/hero-image.png)
![Source](https://en.wikiversity.org/wiki/Internet_Protocol_Analysis/Transport_Layer#/media/File:Internet_Protocol_Analysis_-_Transport_Layer.png)

### Introduction

Each TCP/IP application typically choose to establish a temporary communication session to deliver data between two devices using either Transport Control Protocol (TCP) or User Datagram Protocol (UDP) based on the application requirements; this is done irrespective  regard to the destination host type, the type of media over which the data is to be sent, and the pathway taken by the information [Cisco](https://www.flashcardmachine.com/cisco-1-chapter4.html)
TCP provides a wide range of services such as error recovery, windowing, and order segment delivery to applications, whereas UDP does not.
In this article, we intend to look at these two transport layer protocols in detail.  

### Prerequisites

> To better understand this article, foundational knowledge of TCP/IP  Layer 1 to 3 is required.  

Reviewer: ```suggestion
As a prerequisite, a foundational knowledge of TCP/IP Layer 1 to 3 would help the readers understand the article better.
```

> **Content**
>   -Transport Layer responsibilities  
>   -Multiplexing
>   -Reliability of the transport layer
>   -Transport Control Protocol (TCP)
>   -User Datagram Protocol (UDP)
>   -Multiple conversations
>   -Port number
>   -Socket and Socket pairs
>   -Conclusion

Reviewer: ```suggestion
### Table of contents
- Transport Layer responsibilities  
- Multiplexing
- Reliability of the transport layer
- Transport Control Protocol (TCP)
- User Datagram Protocol (UDP)
- Multiple conversations
- Port number
- Socket and Socket pairs
- Conclusion
```
Suggestions:
- For headings and subheadings, always use markdown headers tags like "###" or "####". Avoid "**".
- For bullet points, use a '-' with a space next to it. Something like "- point 1"
- Do not indent the sentences throughout the article.
- If possible, add the "Table of contents" something like [this](https://www.section.io/engineering-education/data-mining-using-crisp-dm-methodology/#table-of-contents).

> ### Transport Layer responsibility

Reviewer: Before starting with this section, add another section "Transport layer" to explain transport layer in detail and how it works.

>  The Transport layer  has a set of responsibilities to help applications function efficiently and adequately; these responsibilities are:
> -The tracking of individual conversation  
> -Ordered data transfer and data segmentation.
> -Identifying applications.

Reviewer: ```suggestion
The Transport layer has a set of responsibilities to help applications function efficiently and adequately. These responsibilities are:
- The tracking of individual conversation  
- Ordered data transfer and data segmentation.
- Identifying applications.
```

> **Tracking of Individual applications**

Reviewer: Change the subheading to markdown header `####`

Data flowing from one application to another is known as conversation.

>  Because a host can have multiple applications communicating with other applications either within the local network or with a remote network, the transport layer has a mechanism that makes it possible for each application on a host to communicate with another application on a different host either within the local network or in a remote network. According to [Cisco](https://community.cisco.com/t5/switching/which-layer-of-the-model-osi-and-tcp-ip-is-the-port-number/td-p/3356063 ), this mechanism assigns an identifier called a **port number** to each application such that each software process that needs to access the network is has a unique identifier.

Reviewer: ```suggestion
Because a host can have multiple applications communicating with other applications either within the local network or with a remote network, the transport layer has a mechanism that makes it possible for each application on a host to communicate with another application on a different host either within the local network or in a remote network. According to [Cisco](https://community.cisco.com/t5/switching/which-layer-of-the-model-osi-and-tcp-ip-is-the-port-number/td-p/3356063 ), this mechanism assigns an identifier called a **port number** to each application such that each software process that needs to access the network is has a unique identifier.
```

> **Ordered data transfer**  

Reviewer: Change to markdown header `####`

> A continuous stream of bytes is broken down into segments for transmission and delivery by the transport layer's services; this is because, according to [waksudir](http://waksudir.blogspot.com/2015/03/transport-layer-protocols), most networks have a limitation on the amount of data that a single packet can contain.

Reviewer: ```suggestion
A continuous stream of bytes is broken down into segments for transmission and delivery by the transport layer's services. According to [this](http://waksudir.blogspot.com/2015/03/transport-layer-protocols) article, most networks have a limitation on the amount of data that a single packet can contain.
```

>  Because of this, the sending device transport layer prepares the data in manageable pieces or **segments**.

Reviewer: ```suggestion
Because of this, the sending device transport layer prepares the data in manageable pieces or **segments**.
```

>  The receiving device transport layer receives these segments and uses the header to reconstruct them into complete data useful to the application.

Reviewer: ```suggestion
Similarly, the receiving device transport layer receives these segments and uses the header to reconstruct them into complete data useful to the application.
```
Combine this line with the previous line.

> 

Reviewer: Did you miss `Identifying applications`?

Author: I covered it in the first subheading, #### The tracking of individual conversation. 
Sorry I should have deleted it.

### Multiplexing Conversation using port numbers

When using an application, the data or services provided usually appear as a stream of continuous data. 

>   But sending data ( e.g., video) across a network as a complete stream can consume all of the available network bandwidth, thus preventing other communications such as an email from running simultaneously and making error recovery and retransmission of damage data difficult, if not impossible.   

Reviewer: ```suggestion
But sending data ( e.g., video) across a network as a complete stream can consume all of the available network bandwidth, thus preventing other communications such as an email from running simultaneously and making error recovery and retransmission of damage data difficult, if not impossible.   
```

> The Multiplexing mechanism segments TCP and UDP data into small chunks to enable communication from different users to interleave on the same network. This mechanism relies on a concept known as a **socket.**

Reviewer: ```suggestion
The multiplexing mechanism segments TCP and UDP data into small chunks to enable communication from different users to interleave on the same network. This mechanism relies on a concept known as a **socket.**
```

### Transmission Control Protocol

The type of transport layer protocol an application chooses to use depends on the application requirement. 

>  TCP is analogous to sending a package with a tracker that tracks the package from its source to its destination.

Reviewer: ```suggestion
TCP is analogous to sending a package with a tracker that tracks the package from its source to its destination.
```

![TCP Header](/engineering-education/understanding-tcp-ip-transport-layer-protocols/tcp.png)

> ![Source](https://en.wikiversity.org/wiki/Internet_Protocol_Analysis/Transport_Layer#/media/File:Internet_Protocol_Analysis_-_Transport_Layer.png)

Reviewer: ```suggestion
[Source](https://en.wikiversity.org/wiki/Internet_Protocol_Analysis/Transport_Layer#/media/File:Internet_Protocol_Analysis_-_Transport_Layer.png)
```
In markdown, you use `!` only for rendering images.

> As defined in Request For Comment (RFC) 7913, TCP has the following features.
> -Connection establishment
> -Multiplexing using ports
> -Flow control using windowing  
> -Error recovery
> -Ordered data transfer and data segmentation.

Reviewer: ```suggestion
As defined in Request For Comment (RFC) 7913, TCP has the following features:
- Connection establishment
- Multiplexing using ports
- Flow control using windowing  
- Error recovery
- Ordered data transfer and data segmentation.
```

> **Connection establishment and Termination**

Reviewer: Change markdown header

Before any TCP feature can occur, TCP connection establishment must first take place. TCP is a connection-oriented protocol; these protocols establish a permanent connection between client and server before the transfer of data can begin. 

>   During this connection establishment, a device negotiates the amount of traffic to be forwarded during the  Three-way handshake, which must complete before data transfer can begin.

Reviewer: ```suggestion
During this connection establishment, a device negotiates the amount of traffic to be forwarded during the Three-way handshake, which must complete before data transfer can begin.
```

>  For example, to create a three-way handshake between a server and a host, the host sends an SYN flag to a server providing all the necessary information such as its port number (source port) and destination port number (signifying which services it wants access to). When the server receives the SYN flag from the host,  it sends back an SYN and ACK flag, which also contains a source port number (the port number used as the destination port number on the SYN flag sent by the host) and a destination port number (the port number which the host used as source port number). The host acknowledges those flags' reception with an ACK flag, and a connection is established, thus forming a **three-way handshake. **

Reviewer: ```suggestion
For example, to create a three-way handshake between a server and a host, the host sends an SYN flag to a server providing all the necessary information such as its port number (source port) and destination port number (signifying which services it wants access to). When the server receives the SYN flag from the host,  it sends back an SYN and ACK flag, which also contains a source port number (the port number used as the destination port number on the SYN flag sent by the host) and a destination port number (the port number which the host used as source port number). The host acknowledges those flags' reception with an ACK flag, and a connection is established, thus forming a **three-way handshake. **
```

![three-way hanshake](/engineering-education/understanding-tcp-ip-transport-layer-protocols/handshake-image.png)

> ![Source ](Author)

Reviewer: Remove `!`

>  **NOTE:** Each flag must contain a source port number and destination port number. 
>   When a device is sending an SYN flag, its port number becomes the source port number, while that of the device it is sending the flag to becomes the destination port number. After the device receives the SYN flag and wants to send an acknowledgment, it uses the ACK flag and then reverses these port numbers.

Reviewer: ```suggestion
**NOTE:** Each flag must contain a source port number and destination port number. 
When a device is sending an SYN flag, its port number becomes the source port number, while that of the device it is sending the flag to becomes the destination port number. After the device receives the SYN flag and wants to send an acknowledgment, it uses the ACK flag and then reverses these port numbers.
```

The session uses a four-way termination sequence; an additional flag, the FINISHED (FIN) flag, is used together with the SYN and ACK flag.

>   Using the session created above, after the host device receives the last packet from the server, it sends an acknowledgment (ACK) flag informing the server that it has received the packet. 

Reviewer: ```suggestion
Using the session created above, after the host device receives the last packet from the server, it sends an acknowledgment (ACK) flag informing the server that it has received the packet. 
```

>  If the host intends to terminate the session, it sends a FIN flag and the ACK flag, informing the server that it has received all the information it requires from the server and intends to terminate it.
>  The server replies with an ACK flag notifying the client that it has received the FIN flag and is aware of the host's readiness to terminate the session.
>   The server replies with ACK and FIN flags, informing the host of its willingness to end the session. The session ends immediately after the host sends an ACK flag to the server completing the four way-handshake.

Reviewer: ```suggestion
If the host intends to terminate the session, it sends a FIN flag and the ACK flag, informing the server that it has received all the information it requires from the server and intends to terminate it.
The server replies with an ACK flag notifying the client that it has received the FIN flag and is aware of the host's readiness to terminate the session.
The server replies with ACK and FIN flags, informing the host of its willingness to end the session. The session ends immediately after the host sends an ACK flag to the server completing the four way-handshake.
```

![Session termination](/engineering-education/understanding-tcp-ip-transport-layer-protocols/termination-image.png)

> ![Source ](Author)

Reviewer: Remove `!`

> ![Source ](Author)

Reviewer: Remove `!`

>   In the example above, the host initiates the session termination; in practice, any device can terminate a session.

Reviewer: ```suggestion
In the example above, the host initiates the session termination; in practice, any device can terminate a session.
```

  
**Error recovery**
TCP provides reliable data transfer, which means that all packet sent from a source reaches its destination.   In a situation where an error occurs along the route, TCP uses a mechanism to resend the faulty segment.     TCP uses acknowledgment (ACK) and sequence fields in the TCP header to number data bytes and tracks them. By so doing, it achieves reliability.  

> For example, if a web server has 400 bytes of data to send to a web client requesting such data, after establishing the session, the server breakdown the data into small size or segment, let's say, 100 bytes. The server will send the first 100 bytes (0-99) of the data, with a sequence number of 1, the host after receiving this first segment, sends an ACK informing the server that the packet has reached its destination and it is reading to receive the next hundred bytes, i.e., 100-199 with a sequence number of 2. This mechanism will continue up to 400 bytes, with the host sending an acknowledgment for each segment received.

Reviewer: - Change subheading to `####`
- Remove unnecessary spaces

 But with UDP, if a packet is missing, the streaming will continue unnoticed. It only becomes apparent when many segments are missing, which is seen in low video quality and lack of synchronization between video and audio. 
The significant difference between TCP and UDP is that TCP offers a wide range of services to applications, while UDP does not, this does not make UDP inferior to TCP, but by providing fewer services, UDP has fewer bytes in its header, and this makes UDP is faster in transporting data.
![UDP header](/engineering-education/understanding-tcp-ip-transport-layer-protocols/udp-image.png)

> ![Source](https://en.wikiversity.org/wiki/Internet_Protocol_Analysis/Transport_Layer#/media/File:Internet_Protocol_Analysis_-_Transport_Layer.png)

Reviewer: Remove `!`

Reviewer: @raboba2re
Just a minor thing to note.
In markdown, for rendering images you have to add an `!`, and for sources you need not. In the latest fixes, I saw you had removed `!` from both the places, which is incorrect.

Have a look at my commits to understand.

But with UDP, if a packet is missing, the streaming will continue unnoticed. It only becomes apparent when many segments are missing, which is seen in low video quality and lack of synchronization between video and audio.
The significant difference between TCP and UDP is that TCP offers a wide range of services to applications, while UDP does not, this does not make UDP inferior to TCP, but by providing fewer services, UDP has fewer bytes in its header, and this makes UDP is faster in transporting data.

> ![UDP header](/engineering-education/understanding-tcp-ip-transport-layer-protocols/udp-image.png)

Reviewer: ```suggestion
![UDP header](/engineering-education/understanding-tcp-ip-transport-layer-protocols/udp.png)
```
